### PR TITLE
add replacement cycling

### PR DESCRIPTION
**Pull request by mplsgrant**
# Draft
Want to make a CI test for this; waiting on #384 

# Issue
I can not run ariard's [replacement cycling test](https://github.com/bitcoin/bitcoin/commit/9ca92dbfccdc129a467da63f3398714ce06d7b5c#diff-d80e7c9dd2873bbb8e25c5161ca550dcea94da862229c475d9285ab0f07f0f23) in Warnet.

# Cause
A number of the test framework components available in Bitcoin Core have not been ported into Warnet.

# Solution
Add the replacement cycling test as a scenario.

# Wins
The test runs without meaningful augmentation. This is great because it points us in a direction of copy/pasting tests that run today in Core's functional test framework. We can then augment the tests to work in realistic environments with lightning nodes and other future tank types. 

# Depends on
#371

### CODE REVIEW

File: src/scenarios/replacement-cycling.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
+#!/usr/bin/env python3
```
Comment (pinheadmz) [inline src/scenarios/replacement-cycling.py]: file needs to be named with underscore not hyphen. rpc_test.py fails locally for me without that fix, dunno why CI is ok with it 🤷

**Review by pinheadmz — COMMENTED** [review]
@mplsgrant are you able to get this scenario running in warnet? I got a connect_nodes error at first then rebased on main, now I get a sync_all error:

```
│  2024-06-07T14:32:33+0000 - INFO - Traceback (most recent call last):                                                                                                        │
│  2024-06-07T14:32:33+0000 - INFO -   File "/root/warnet/src/test_framework/test_framework.py", line 131, in main                                                             │
│  2024-06-07T14:32:33+0000 - INFO -     self.run_test()                                                                                                                       │
│  2024-06-07T14:32:33+0000 - INFO -   File "/root/warnet/src/scenarios/replacement_cycling.py", line 430, in run_test                                                         │
│  2024-06-07T14:32:33+0000 - INFO -     self.generatetoaddress(self.nodes[0], nblocks=101,                                                                                    │
│  2024-06-07T14:32:33+0000 - INFO -   File "/root/warnet/src/test_framework/test_framework.py", line 692, in generatetoaddress                                                │
│  2024-06-07T14:32:33+0000 - INFO -     sync_fun() if sync_fun else self.sync_all()                                                                                           │
│  2024-06-07T14:32:33+0000 - INFO -                                 ^^^^^^^^^^^^^^^                                                                                           │
│  2024-06-07T14:32:33+0000 - INFO -   File "/root/warnet/src/test_framework/test_framework.py", line 746, in sync_all                                                         │
│  2024-06-07T14:32:33+0000 - INFO -     self.sync_blocks(nodes)                                                                                                               │
│  2024-06-07T14:32:33+0000 - INFO -   File "/root/warnet/src/test_framework/test_framework.py", line 715, in sync_blocks                                                      │
│  2024-06-07T14:32:33+0000 - INFO -     assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
```

File: src/scenarios/replacement-cycling.py (hunk 1 above)
Comment (mplsgrant) [reply src/scenarios/replacement-cycling.py]: I thought the scenarios test would have caught that.

**Comment by mplsgrant** [comment]
> are you able to get this scenario running in warnet?

Yes, we got a version of it working just fine [on this branch](https://github.com/mplsgrant/warnet/tree/2024-05-28-instagibbs-demo). There shouldn't be any differences that `sync_all` would have trouble with. I'll rebase and have a look.

**Comment by mplsgrant** [comment]
I rebased this.

**Comment by mplsgrant** [comment]
Rebased on main

**Review by willcl-ark — COMMENTED** [review]
Can we add a CI test for this? I suspect it may fail, as I find locally:

```log
 2024-06-21T10:29:59+0000 - INFO - Started server
 * Serving Flask app 'warnet.server'
 * Debug mode: on
 2024-06-21T10:30:00+0000 - INFO - Logging started
 2024-06-21T10:30:00+0000 - INFO - Started server
 2024-06-21T10:30:00+0000 - WARNING -  * Debugger is active!
 2024-06-21T10:30:06+0000 - DEBUG - {'jsonrpc': '2.0', 'method': 'scenarios_run', 'params': {'scenario': 'replacement_cycling', 'additional_args': [], 'network': 'warnet'}, 'id': 1}
 2024-06-21T10:30:06+0000 - DEBUG - Running ['/usr/local/bin/python3', '/root/warnet/src/scenarios/replacement_cycling.py', '--network=warnet']
 2024-06-21T10:30:06+0000 - INFO - Unexpected exception caught during testing
 2024-06-21T10:30:06+0000 - INFO - Traceback (most recent call last):
 2024-06-21T10:30:06+0000 - INFO -   File "/root/warnet/src/test_framework/test_framework.py", line 130, in main
 2024-06-21T10:30:06+0000 - INFO -     self.setup()
 2024-06-21T10:30:06+0000 - INFO -   File "/root/warnet/src/warnet/test_framework_bridge.py", line 60, in setup
 2024-06-21T10:30:06+0000 - INFO -     self.warnet = Warnet.from_network(self.options.network)
 2024-06-21T10:30:06+0000 - INFO -                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 2024-06-21T10:30:06+0000 - INFO -   File "/root/warnet/src/warnet/warnet.py", line 137, in from_network
 2024-06-21T10:30:06+0000 - INFO -     self.graph = networkx.read_graphml(
 2024-06-21T10:30:06+0000 - INFO -                  ^^^^^^^^^^^^^^^^^^^^^^
 2024-06-21T10:30:06+0000 - INFO -   File "/usr/local/lib/python3.12/site-packages/networkx/utils/decorators.py", line 770, in func
 2024-06-21T10:30:06+0000 - INFO -     return argmap._lazy_compile(__wrapper)(*args, **kwargs)
 2024-06-21T10:30:06+0000 - INFO -            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 2024-06-21T10:30:06+0000 - INFO -   File "<class 'networkx.utils.decorators.argmap'> compilation 5", line 3, in argmap_read_graphml_1
 2024-06-21T10:30:06+0000 - INFO -     import gzip
 2024-06-21T10:30:06+0000 - INFO -
 2024-06-21T10:30:06+0000 - INFO -   File "/usr/local/lib/python3.12/site-packages/networkx/utils/decorators.py", line 193, in _open_file
 2024-06-21T10:30:06+0000 - INFO -     fobj = _dispatch_dict[ext](path, mode=mode)
 2024-06-21T10:30:06+0000 - INFO -            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 2024-06-21T10:30:06+0000 - INFO - FileNotFoundError: [Errno 2] No such file or directory: '/root/.warnet/warnet/warnet/graph.graphml'

```

I can get it to run, but only if I run a manual warnet first. This I presume creates the config dir and copies the graph into it which this test is looking for.

Perhaps this test should be explicit about which graph it is using as is done in the [other tests](https://github.com/bitcoin-dev-project/warnet/blob/dedf2054b188c91b86ff0912904242e24d7fec25/test/scenarios_test.py#L8-L12) which would load and copy the graph correctly?

**Comment by mplsgrant** [comment]
@willcl-ark Happy to add a CI for this. 

When the scenario failed locally, were these the steps you took?

```shell
just start
just p &
warcli scenarios run replacement_cycling
```

**Comment by mplsgrant** [comment]
Closing in favor of https://github.com/bitcoin-dev-project/warnet/pull/422

**Pull request closed without merging**